test/utility.coffee

```diff
@@ -0,0 +1,82 @@
+###
+Copyright (C) 2016 Thomas Rosenau
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+
+  * Redistributions of source code must retain the above copyright
+    notice, this list of conditions and the following disclaimer.
+  * Redistributions in binary form must reproduce the above copyright
+    notice, this list of conditions and the following disclaimer in the
+    documentation and/or other materials provided with the distribution.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 'AS IS'
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
+DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+###
+
+'use strict'
+
+utility = require '../lib/utility'
+expect = require('chai').expect
+
+describe 'utility:', ->
+    describe 'deepCopy:', ->
+        it 'copies objects', ->
+            o = {}
+            copy = utility.deepCopy o
+            expect(JSON.stringify copy).to.equal JSON.stringify o
+            expect(copy).not.to.equal o
+
+            a = []
+            copy = utility.deepCopy a
+            expect(JSON.stringify copy).to.equal JSON.stringify a
+            expect(copy).not.to.equal a
+
+            o2 = {a: 1, b: 2, c: [{d: true}]}
+            copy = utility.deepCopy o2
+            expect(JSON.stringify copy).to.equal JSON.stringify o2
+            expect(copy).not.to.equal o2
+            expect(copy.c).not.to.equal o2.c
+            expect(copy.c[0]).not.to.equal o2.c[0]
+
+        it 'filters internal properties', ->
+            o = {__foo: true, bar: [{__baz: 1}, 2]}
+            copy = utility.deepCopy o
+            expect(copy.__foo).to.equal undefined
+            expect(copy.bar[0].__baz).to.equal undefined
+            expect(JSON.stringify copy).to.equal JSON.stringify {bar: [{}, 2]}
+
+    describe 'generateNextName:', ->
+        it 'generates names', ->
+            #XXX not valid identifiers - should we disallow these input values by throwing an error?
```

Comment:
No, do not disallow them, but it should be undefined behaviour, so remove the next 4 tests.